### PR TITLE
RichText: Fix empty value delete behaviors

### LINKED ISSUE #8731: Writing Flow: Can't forward delete from empty paragraph
**To Reproduce**

1. Navigate to Posts > Add New
2. Click writing prompt
3. Press Enter
4. Press ArrowUp
5. Press Delete (Control + Delete on Mac) for forward deletion
6. Observe nothing happens

**Expected behavior**

The next paragraph should collapse into the current.

### PR DESCRIPTION
**Pull request by aduth**
Fixes #8731 

This pull request seeks to resolve two issues related to deletion behaviors on empty RichText (paragraph blocks). The first is described in #8731, the second I'm not sure has yet been surfaced (and exists in the production v3.4 release):

1. Navigate to Posts > Add New
2. Click writing prompt
3. Press Enter twice
4. Click first paragraph
5. Press (forward) Delete
6. Observe two paragraph deletion instead of one

The latter of these is due to the fact that we did not distinguish between `onMerge` and `onRemove` in the key handlers of RichText, thus causing _both_ to fire, merging into the one block and then deleting that one block, resulting in the double-deletion. This has been accounted for here by updating the block's `onMerge` to ~return a boolean indicating whether it was successful in its merge. This is interpreted by `RichText` to determine whether a removal is warranted, if possible~ only handle removal on the Backspace key (see https://github.com/WordPress/gutenberg/pull/8735#issuecomment-411776551).

The issue causing #8731 is a result of the fact that when focused in an empty RichText, the caret is placed adjacent (before) a "bogus" `br` node inserted by TinyMCE. The implementation of the general purpose `isHorizontalEdge` interprets this to mean that the caret is not at the end of the field, thus it would never allow the removal.

**Implementation notes:**

I adopted `waitForRichTextInitialization` from #8306 because I was having issues with failures locally without it. I'm not terribly happy with its inclusion, also because it only works for explicit inserts of the paragraph block, not a new paragraph inserted via Enter or other keyboard interactions.

**Testing instructions:**

Repeat testing instructions from #8731 and the above second set of issue steps, verifying that paragraph merges as expected, with a maximum of one removal.

Ensure end-to-end tests pass:

```
npm run test-e2e
```

### CODE REVIEW

**Comment by aduth** [comment]
The failing end-to-end test is legitimate, and indicates that we rely on RichText to remove itself even when calling `onMerge` and a block does exist in the direction of the delete (before / after), but it's not mergeable. For example, an empty paragraph following an image block, when pressing Backspace, will _both_ merge and remove (in master), and thus fails with the changes as proposed here.

We seem to be quite inconsistent in how we expect this to work. It doesn't seem like we should want both a merge and remove to happen, but rather one or the other, depending on whether a merge is possible and if the current RichText is empty.

Some ideas off the top of my head:

- Somehow account for "mergeable blocks" in the newly-introduced boolean return value of `onMerge` here
- When RichText is empty and delete is pressed, don't try to merge at all, just remove, and somehow trigger the appropriate caret movement to the next / previous block

**Review by tofumatt — COMMENTED** [review]
The code here makes sense but I'm slightly too tired to offer much more input on 
https://github.com/WordPress/gutenberg/pull/8735#issuecomment-411585846

I'll just leave my comments here. 🤷‍♂️

File: packages/editor/src/components/rich-text/index.js [hunk 1]
```diff
@@ -426,41 +427,84 @@ export class RichText extends Component {
...
-		if (
-			getSelection().isCollapsed && (
-				( keyCode === BACKSPACE && isHorizontalEdge( rootNode, true ) ) ||
-				( keyCode === DELETE && isHorizontalEdge( rootNode, false ) )
-			)
-		) {
-			if ( ! this.props.onMerge && ! this.props.onRemove ) {
-				return;
-			}
+		// Only process delete if occurs at uncollapsed edge.
```
[the remark this answers is not in the document]
Comment (youknowriad) [reply packages/editor/src/components/rich-text/index.js]: I think it's the opposite: 

`// Only process delete if the deletion occurs at uncollapsed edge.`

File: packages/editor/src/components/rich-text/README.md [hunk 2]
```diff
@@ -41,11 +41,11 @@ Render a rich [`contenteditable` input](https://developer.mozilla.org/en-US/docs
 
 ### `onMerge( forward: Boolean ): Function`
 
-*Optional.* Called when blocks can be merged. `forward` is true when merging with the next block, false when merging with the previous block.
+*Optional.* Called when blocks can be merged. `forward` is true when merging with the next block, false when merging with the previous block. The callback should return `true` if a merge is possible, or `false` otherwise. This is used in determining whether an empty field should be deleted instead of merged.
```
[the remark this answers is not in the document]
Comment (aduth) [reply packages/editor/src/components/rich-text/README.md]: Changes backed out in rebased d29d5d1.

File: packages/editor/src/components/rich-text/README.md [hunk 3]
```diff
@@ -41,11 +41,11 @@ Render a rich [`contenteditable` input](https://developer.mozilla.org/en-US/docs
...
 
 ### `onRemove( forward: Boolean ): Function`
 
-*Optional.* Called when the block can be removed. `forward` is true when the selection is expected to move to the next block, false to the previous block.
+*Optional.* Called when the block can be removed. `forward` is true when the selection is expected to move to the next block, false to the previous block. Called when pressing Backspace or Delete on an empty field and only when `onMerge` explicitly returns `false` to indicate non-merge.
```
[the remark this answers is not in the document]
Comment (aduth) [reply packages/editor/src/components/rich-text/README.md]: Changes backed out in rebased d29d5d1.

File: packages/editor/src/components/rich-text/index.js [hunk 4]
```diff
@@ -426,41 +427,84 @@ export class RichText extends Component {
 	}
 
 	/**
-	 * Handles a keydown event from TinyMCE.
+	 * Handles a delete key down event to handle merge or removal for a
```
[the remark this answers is not in the document]
Comment (aduth) [reply packages/editor/src/components/rich-text/index.js]: Updated in rebased d29d5d1.

File: packages/editor/src/components/rich-text/index.js [hunk 5]
```diff
@@ -426,41 +427,84 @@ export class RichText extends Component {
...
+	 * collapsed selection where caret is at directional edge: forward for a
```
[the remark this answers is not in the document]
Comment (aduth) [reply packages/editor/src/components/rich-text/index.js]: Added link in rebased d29d5d1.

File: packages/editor/src/components/rich-text/index.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (aduth) [reply packages/editor/src/components/rich-text/index.js]: Updated in rebased d29d5d1.

**Comment by aduth** [comment]
Okay, the latest rebased iteration seems to handle all cases. The major change is only handling `onRemove` from `RichText` if the key pressed is the backspace key. This seems like the more correct interaction to expect for removal anyways, in my opinion.

**Review by tofumatt — APPROVED** [review]
I'm unclear about:

> with a maximum of one removal

in the testing instructions. Does that mean the forward delete key shouldn't trigger multiple empty paragraph deletions? It seems like it should conceptually, and it does:



This works great though, otherwise tested well locally, felt good to use, and tests passed locally for me. I made a tiny tweak (`keydown` to `keyDown` in a comment because I thought it made it more readable as an event).

🚢

**Comment by aduth** [comment]
>Does that mean the forward delete key shouldn't trigger multiple empty paragraph deletions?

Yep, for example, in the 3.4 release if you had multiple empty paragraphs, hitting (forward) Delete in the first of them would destroy _two_.

In the following recording, I hit Delete once, but two paragraphs are deleted (and the caret randomly moves to the preceding paragraph):

forward-delete

**Comment by tofumatt** [comment]
Oh wait, okay, I understand the instruction now. I just misinterpreted what you meant. Aces! 👍

**Pull request merged**